### PR TITLE
Add patch syntax.

### PR DESCRIPTION
**Pull request by scobbe**

### CODE REVIEW

File: stone/frontend/ir_generator.py [hunk 1]
```diff
@@ -319,6 +328,73 @@ def _create_type(self, env, item):
...
+            patched_item_base_name = self._get_base_name(patched_item.name, patched_namespace.name)
+            if patched_item_base_name not in self._item_by_canonical_name:
+                raise InvalidSpec('Patch {} must correspond to a pre-existing data_type.'.format(
+                    quote(patched_item.name)), patched_item.lineno, patched_item.path)
+
+            existing_item = self._item_by_canonical_name[patched_item_base_name]
+
+            self._check_patch_type_mismatch(patched_item, existing_item)
+
+            if (isinstance(patched_item, AstStructPatch)
```
Comment (braincore) [inline stone/frontend/ir_generator.py]: Can compress it into: `isinstance(patched_item, (AstStructPatch, AstUnionPatch))`

File: stone/frontend/ir_generator.py [hunk 2]
```diff
@@ -319,6 +328,73 @@ def _create_type(self, env, item):
...
+                quote(patched_item.name),
+                quote(existing_item.name),
+                existing_item.path,
+                existing_item.lineno,
+                quote(data_type_name)), patched_item.lineno, patched_item.path)
+
+        if isinstance(patched_item, AstStructPatch):
+            if not isinstance(existing_item, AstStructDef):
+                raise_mismatch_error(patched_item, existing_item, 'struct')
+        elif isinstance(patched_item, AstUnionPatch):
```
Comment (braincore) [inline stone/frontend/ir_generator.py]: Enforce that the original union data type and union patched data type have a matching close policy. Either both closed or open.

File: test/test_stone.py [hunk 3]
```diff
@@ -1315,6 +1445,340 @@ def test_forward_reference(self):
...
+            """)
+        api = specs_to_ir([('test.stone', text)])
+
+        quota_info_dt = api.namespaces['files'].data_type_by_name['QuotaInfoPersonal']
+        self.assertEqual(quota_info_dt.all_fields[1].name, 'quota')
+        self.assertTrue(is_integer_type(quota_info_dt.all_fields[1].data_type))
+        self.assertEqual(
+            quota_info_dt.all_fields[1].doc, "The user's total quota allocation (bytes).")
+
+        # Test patching enumerated supertype struct
```
Comment (braincore) [inline test/test_stone.py]: Testing patching the parent type of an enumerated subtype.

File: test/test_stone.py [hunk 4]
```diff
@@ -1315,6 +1445,340 @@ def test_forward_reference(self):
         self.assertEqual(
             api.namespaces['test'].data_types[0].fields[0].default.tag_name, 'a')
 
+    def test_struct_patch_semantics(self):
```
Comment (braincore) [inline test/test_stone.py]: How about a test patching a struct that hasn't been defined?

File: test/test_stone.py [hunk 5]
```diff
@@ -1315,6 +1445,340 @@ def test_forward_reference(self):
...
+                size UInt64
+                    "The size of the file."
+                example default
+                    size=5
+            struct Folder extends Resource
+                icon String
+                    "The name of the icon."
+                example default
+                    icon="My Icon"
+            patch struct Resource
```
Comment (braincore) [inline test/test_stone.py]: What if I tried to declare a new union subtype here?

File: test/test_stone.py [hunk 6]
```diff
@@ -1315,6 +1445,340 @@ def test_forward_reference(self):
...
+                    user_id="1234"
+            """)
+        with self.assertRaises(InvalidSpec) as cm:
+            api = specs_to_ir([('test.stone', text)])
+
+        self.assertEqual("Example defined in patch 'QuotaInfo' must " +
+            "correspond to a pre-existing example.", cm.exception.msg)
+        self.assertEqual(cm.exception.lineno, 13)
+
+    def test_union_patch_semantics(self):
```
Comment (braincore) [inline test/test_stone.py]: Make sure you have a test for patching an open union with a closed union and vice versa.

File: docs/lang_ref.rst [hunk 7]
```diff
@@ -636,6 +636,75 @@ is prohibited to make generating languages like Python possible.
...
+Now in ``private/people.stone``, we can define additional fields::
+
+    namespace people
+
+    patch struct Person
+        age UInt64
+            "The number of years, rounded down."
+
+Only data types that have been fully-defined elsewhere can be patched. Note that patching
+can only be used to add additional fields, not mutate existing fields.
```
[the remark this answers is not in the document]
Comment (scobbe) [reply docs/lang_ref.rst]: Spoke offline, will leave as is.

File: stone/frontend/ir_generator.py (hunk 2 above)
Comment (scobbe) [reply stone/frontend/ir_generator.py]: Good catch, I added a check / corresponding test.

File: test/test_stone.py (hunk 4 above)
Comment (scobbe) [reply test/test_stone.py]: This exists already, look for comment `# Try patching type without pre-existing type`

File: docs/lang_ref.rst [hunk 8]
```diff
@@ -636,6 +636,75 @@ is prohibited to make generating languages like Python possible.
...
+
+Patching can inject both required and optional fields, but in the required fields case,
```
[the remark this answers is not in the document]
Comment (scobbe) [reply docs/lang_ref.rst]: Done.

File: docs/lang_ref.rst [hunk 9]
```diff
@@ -636,6 +636,75 @@ is prohibited to make generating languages like Python possible.
...
+    struct Person
+        "Describes a member of society."
+
+        name String
+            "Given name followed by surname."
+
+        example default
+            name = "Anonymous Person"
+
+        example child
```
[the remark this answers is not in the document]
Comment (scobbe) [reply docs/lang_ref.rst]: Done. You're listed as a child.

File: stone/frontend/ir_generator.py (hunk 1 above)
Comment (scobbe) [reply stone/frontend/ir_generator.py]: Done.

File: test/test_stone.py (hunk 3 above)
Comment (scobbe) [reply test/test_stone.py]: Good idea. Done.

File: test/test_stone.py (hunk 6 above)
Comment (scobbe) [reply test/test_stone.py]: Done.

File: test/test_stone.py (hunk 3 above)
Comment (scobbe) [reply test/test_stone.py]: Done.

File: test/test_stone.py (hunk 5 above)
Comment (scobbe) [reply test/test_stone.py]: This fails in the parsing layer, albeit with a rather unclear message. I added a test. Look for `# test no patching enumerated subtype`.

**Comment by scobbe** [comment]
Okay, I responded to your feedback, @braincore!

**Comment by braincore** [comment]
Last thing: have a meatier commit message.

**Comment by braincore** [comment]
When you create the annotations diff, be sure to update `docs/managing_specs.rst` as well.

**Pull request merged**